pilot/pkg/model/service.go

```diff
@@ -1080,34 +1080,55 @@ func IsDNSSrvSubsetKey(s string) bool {
 	return false
 }
 
+// ParseSubsetKeyHostname is an optimized specialization of ParseSubsetKey that only returns the hostname.
+// This is created as this is used in some hot paths and is about 2x faster than ParseSubsetKey; for typical use ParseSubsetKey is sufficient (and zero-alloc).
+func ParseSubsetKeyHostname(s string) (hostname string) {
+	idx := strings.LastIndex(s, "|")
+	if idx == -1 {
+		// Could be DNS SRV format.
+		// Do not do LastIndex("_."), as those are valid characters in the hostname (unlike |)
+		// Fallback to the full parser.
```

Comment:
why not call the similar function with `_.`

Reply:
It's in the comment; _. can be in the hostname so finding the last _. may not give us the full hostname. Go doesn't have  a function to find the Nth index.

The fall back is still pretty fast anyways so not that big of a concern IMO. There are usually less than 1% of calls for this cluster name format